Mathlib/Topology/Order/UpperSetTopology.lean

```diff
@@ -0,0 +1,283 @@
+  right_inv := toUpperSet_ofUpperSet,
+  map_rel_iff' := toUpperSet_le_iff }
+
+end WithUpperSetTopology
+
+/--
+The upper set topology is the topology where the open sets are the upper sets. In general the upper
+set topology does not coincide with the dual of the lower topology.
+-/
+class UpperSetTopology (α : Type _) [t : TopologicalSpace α] [Preorder α] : Prop where
+  topology_eq_upperSetTopology : t = upperSetTopology' α
+
+instance [Preorder α] : UpperSetTopology (WithUpperSetTopology α) :=
+  ⟨rfl⟩
+
+instance [Preorder α] : @UpperSetTopology (WithUpperSetTopology α) (upperSetTopology' α) _ := ⟨rfl⟩
+
+instance [Preorder α] : @UpperSetTopology α (upperSetTopology' α) _ := by
+  letI := upperSetTopology' α
+  exact ⟨rfl⟩
+
+
+namespace UpperSetTopology
+
+section Preorder
+
+variable (α)
+variable [Preorder α] [TopologicalSpace α] [UpperSetTopology α] {s : Set α}
+
+lemma topology_eq : ‹_› = upperSetTopology' α := topology_eq_upperSetTopology
+
+variable {α}
+
+/-- If `α` is equipped with the upper set topology, then it is homeomorphic to
+`WithUpperSetTopology α`.
+-/
+def withUpperSetTopologyHomeomorph : WithUpperSetTopology α ≃ₜ α :=
+  WithUpperSetTopology.ofUpperSet.toHomeomorphOfInducing ⟨by erw [topology_eq α, induced_id]; rfl⟩
+
+lemma IsOpen_iff_IsUpperSet : IsOpen s ↔ IsUpperSet s := by
+  rw [topology_eq α]
+  rfl
+
+-- Alexandrov property, set formulation
+theorem IsOpen_sInter {S : Set (Set α)} (hf : ∀ s ∈ S, IsOpen s) : IsOpen (⋂₀ S) := by
+  rw [IsOpen_iff_IsUpperSet]
+  apply isUpperSet_sInter
+  intros s hs
+  rw [← IsOpen_iff_IsUpperSet]
+  exact hf _ hs
+
+-- Alexandrov property, index formulation
+theorem isOpen_iInter {f : ι → Set α} (hf : ∀ i, IsOpen (f i)) : IsOpen (⋂ i, f i) := by
+  rw [IsOpen_iff_IsUpperSet]
+  apply isUpperSet_iInter
+  intros i
+  rw [← IsOpen_iff_IsUpperSet]
+  exact hf i
+
+-- c.f. isClosed_iff_lower_and_subset_implies_LUB_mem
+lemma isClosed_iff_isLower {s : Set α} : IsClosed s
+  ↔ (IsLowerSet s) := by
+  rw [← isOpen_compl_iff, IsOpen_iff_IsUpperSet,
+    isLowerSet_compl.symm, compl_compl]
+
+lemma isClosed_isLower {s : Set α} : IsClosed s → IsLowerSet s := fun h =>
+  (isClosed_iff_isLower.mp h)
+
+lemma closure_eq_lowerClosure {s : Set α} : closure s = lowerClosure s := by
+  rw [subset_antisymm_iff]
+  constructor
+  · apply closure_minimal subset_lowerClosure _
+    rw [isClosed_iff_isLower]
+    exact LowerSet.lower (lowerClosure s)
+  · apply lowerClosure_min subset_closure (isClosed_isLower isClosed_closure)
+
+/--
+The closure of a singleton `{a}` in the upper set topology is the right-closed left-infinite
+interval (-∞,a].
+-/
+@[simp] lemma closure_singleton {a : α} : closure {a} = Iic a := by
+  rw [closure_eq_lowerClosure, lowerClosure_singleton]
+  rfl
+
+end Preorder
+
+section maps
+
+variable [Preorder α] [Preorder β]
+
+lemma upperSetTopology_coinduced {t₁ : TopologicalSpace α} [UpperSetTopology α]
+  (hf : Monotone f) : coinduced f t₁ ≤ upperSetTopology' β := by
```

Comment:
```suggestion
    (hf : Monotone f) : coinduced f t₁ ≤ upperSetTopology' β := by
```
The new convention is that everything before the `:=` has to be indented twice. Could you also check that in the rest of the file?

Reply:
Presumably this only applies when the statement of the result runs over multiple lines? e.g.
```
lemma Monotone_to_LowerTopology_Dual_Continuous [TopologicalSpace α]
  [UpperSetTopology α] [TopologicalSpace β] [LowerTopology β] {f : α → βᵒᵈ} (hf : Monotone f) :
    Continuous f:= by
  rw [continuous_def]
```

Reply:
No from what I understand the hypotheses and the statement have the same level of indentation, you only go back one tab when getting into the proof. So that gives:
```lean
lemma Monotone_to_LowerTopology_Dual_Continuous [TopologicalSpace α]
    [UpperSetTopology α] [TopologicalSpace β] [LowerTopology β] {f : α → βᵒᵈ} (hf : Monotone f) :
    Continuous f := by
  rw [continuous_def]
```

Reply:
Hopefully I've got this correct now?

Reply:
Yes that's perfect, thanks!